### DISCUSSION TITLE
Store user cache in mysql

### DISCUSSION THREAD

**Discussion by 0xIO32** [PurpurMC/Purpur#1341] [__Dismissed__] [answered]
Surely it would be better to store the user cache in mysql. But it should be implemented in such a way that it works with the MySQL Connection and it is not simply loaded into RAM, because it should also work for multiple servers. So a shared usercache.

**Comment by YouHaveTrouble** [comment]
What happens when mysql dies or takes long to respond? There are always tradeoffs for any solution. Sure, using a file loaded in memory only works for a single server, but it's also really safe since it can't really fail in any way. Single server is also by far most common setup and mysql solution doesn't really benefit that and even might be a detriment.

**Reply by 0xIO32** [reply]
Then it could be resolved with the mojang api or simply returned as that the user isn't cached.

But same problem as for example a permission plugin. An there it is implemented.

And for the argument, "Single server is also by far most common setup". Thats why it should be optional.

**Reply by YouHaveTrouble** [reply]
> Then it could be resolved with the mojang api or simply returned as that the user isn't cached.

This would absolutely break plugin api. There is a commonly used method `Bukkit.getOfflinePlayerIfCached()`. This is expected to be extremely fast and safe to use on main thread, but if you introduce a mysql query into the retrieval of the value of this function it suddenly blocks main thread for the time that query takes, which is way too long to justify being safe to use on a main thread. This would lead to many plugins just blocking the main thread and watchdog killing servers because a plugin decided to do loop over some players.

**Reply by YouHaveTrouble** [reply]
Don't get me wrong, it's an interesting idea, but it's not realistic to implement when all things it affects are considered.

**Reply by 0xIO32** [reply]
First of all MySQL Querys are not taking long. If the connections aborts, it will try to reconnect and this will take long.

But i see the problem. Maybe we could replace the save and load functionality with mysql and store it to ram anyway. The cache should only add players and not remove. With a Primary Key we can ensure, that there are not duplicates.

But i am unsure if the function has sense, when it is stored into memory anyway.

**Reply by YouHaveTrouble** [reply]
> First of all MySQL Connections are not taking long.

They are taking long for the time sensitive things like main thread time. There are 50 milliseconds in the tick, literally any remote mysql connection will take at least 5-10ms. Even assuming best case scenario it's too significant.

**Reply by 0xIO32** [reply]
And just loading into memory and updating every 5-10 minutes (configurable) for example?

**Reply by YouHaveTrouble** [reply]
> And just loading into memory and updating every 5-10 minutes (configurable) for example?

Would make `Bukkit.getOfflinePlayerIfCached()` return false values for players that logged in to the server, which is a major api breakage

**Comment by 0xIO32** [comment]
I mean, when someone joins which is not in the cache. Then it will be added in the memory and then there will be an async mysql request to add it. The Primary Key will stop that there will be duplicates. It shouldn't break anything.

**Reply by YouHaveTrouble** [reply]
Even when you make an async request to fill the value, the api call that expects the result instantly will return wrong value. I really considered this feature, really. It seems cool, but it's not implementable without major api breaks

**Reply by 0xIO32** [reply]
I don't unterstand how can it return a wrong value.

on startup -> it loads from the database.
 -> it has all entries

player join -> (it inserts the entry in the database **and directly in the ram too**)
-> it has all player cache entries from this server instance, duplicates does not exist because of the primary key in the database

async scheduler -> safely updates the cache entry in an configurable interval with the mysql database.
-> for player entry updates of the cache from other server instances

Bukkit.getOfflinePlayerIfCached() -> value from **ram** which has all entries from this server instance.

**Comment by katekerllenevich** [comment, marked as answer]
I think that there are plenty of possible issues with something like this that really aren't up for debate. This is also far, _far_, out of scope of Purpur's job. In fact, it sounds exactly like what some plugins do for syncing player data between BungeeCord/Velocity instances. The idea of including a MySQL connection built into the API is nice in a perfect world, but unrealistic at best and feels like something that would come out of the 1.6 era of Minecraft.
